src/MongoDB.Driver/Linq/Linq3Implementation/Serializers/KnownSerializers/KnownSerializerFinder.cs

```diff
@@ -22,33 +27,139 @@ internal class KnownSerializerFinder : ExpressionVisitor
     {
         #region static
         // public static methods
-        public static KnownSerializersRegistry FindKnownSerializers(Expression root)
+        public static KnownSerializersRegistry FindKnownSerializers(Expression root, IBsonDocumentSerializer rootSerializer)
         {
-            var visitor = new KnownSerializerFinder();
+            var visitor = new KnownSerializerFinder(root, rootSerializer);
             visitor.Visit(root);
             return visitor._registry;
         }
         #endregion
 
         // private fields
-        private KnownSerializersNode _expressionKnownSerializers = null;
-        private readonly KnownSerializersRegistry _registry = new KnownSerializersRegistry();
+        private KnownSerializersNode _currentKnownSerializersNode;
+        private IBsonDocumentSerializer _currentSerializer;
+        private readonly KnownSerializersRegistry _registry = new();
+        private readonly Expression _root;
+        private readonly IBsonDocumentSerializer _rootSerializer;
 
         // constructors
-        public KnownSerializerFinder()
+        private KnownSerializerFinder(Expression root, IBsonDocumentSerializer rootSerializer)
         {
+            _rootSerializer = rootSerializer;
+            _root = root;
         }
 
         // public methods
         public override Expression Visit(Expression node)
         {
-            _expressionKnownSerializers = new KnownSerializersNode(_expressionKnownSerializers);
-            _registry.Add(node, _expressionKnownSerializers);
+            if (node == null) return null;
```

Comment:
nit: we avoid one liners like this because things to the right are harder to see.

Preferred:

    if (node == null)
    {
        return null;
    }

Reply:
Fixed.